File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/WipeRestartViewModel.kt
// E4 — Kotlin mirror of FlagshipUI's WipeRestartViewModel. Rotates
// UMK + IRK + recovery passkey + push token in one atomic ceremony.
// See WipeRestartViewModel.swift for the canonical commentary.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.WipeRestartRequest
import com.flagshipserver.app.core.HexUtil
import com.flagshipserver.app.core.WipeRestartClaim
import com.flagshipserver.app.keystore.Keystore
import com.flagshipserver.app.keystore.MockWebAuthnProvider
import com.flagshipserver.app.keystore.Recovery
import com.flagshipserver.app.keystore.WebAuthnProvider
import com.google.crypto.tink.subtle.Ed25519Sign
import com.google.crypto.tink.subtle.Hkdf
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.security.MessageDigest
import java.security.SecureRandom

sealed interface WipeRestartPhase {
    data object Idle : WipeRestartPhase
    data object PreparingKeys : WipeRestartPhase
    data object RegisteringPasskey : WipeRestartPhase
    data object WrappingNewUmk : WipeRestartPhase
    data object Signing : WipeRestartPhase
    data object Posting : WipeRestartPhase
    data object InstallingLocally : WipeRestartPhase
    data object Completed : WipeRestartPhase
    data class Failed(val message: String) : WipeRestartPhase
}

class WipeRestartViewModel(
    private val server: FlagshipServerClient,
    private val webAuthn: WebAuthnProvider = MockWebAuthnProvider(),
    private val username: () -> String?,
    private val rng: SecureRandom = SecureRandom(),
) {
    private val _phase = MutableStateFlow<WipeRestartPhase>(WipeRestartPhase.Idle)
    val phase: StateFlow<WipeRestartPhase> = _phase.asStateFlow()

    suspend fun run(currentEtag: String?) {
        val user = username()
        if (user.isNullOrEmpty()) {
            _phase.value = WipeRestartPhase.Failed("No active account on this device.")
            return
        }
        _phase.value = WipeRestartPhase.PreparingKeys

        // 1 — Read OLD UMK + derive OLD IRK while the existing
        // identity is still active.
        val oldUmk: ByteArray
        val oldIrkSign: Ed25519Sign
        val oldIrkPubHex: String
        try {
            oldUmk = Keystore.currentUmkSeed()
            oldIrkSign = Keystore.deriveIRK("Authorize wipe", version = Keystore.currentIrkVersion())
            // We need the OLD IRK pub for the canonical-bytes claim.
            // Re-derive via the cached seed slot.
            val oldSeed = Keystore.requireIrkSeedForVersion(Keystore.currentIrkVersion())
            val pair = Ed25519Sign.KeyPair.newKeyPairFromSeed(oldSeed)
            oldIrkPubHex = HexUtil.encode(pair.publicKey)
        } catch (e: Throwable) {
            _phase.value = WipeRestartPhase.Failed("Couldn't access your current key: ${e.message}")
            return
        }

        // 2 — Fresh UMK + passkey register.
        val newUmk = ByteArray(32).also(rng::nextBytes)
        _phase.value = WipeRestartPhase.RegisteringPasskey
        val credentialId: String
        try {
            credentialId = webAuthn.register()
        } catch (e: Throwable) {
            _phase.value = WipeRestartPhase.Failed("Passkey registration failed: ${e.message}")
            return
        }
        val prfSecret: ByteArray
        try {
            prfSecret = webAuthn.prfAssert(credentialId)
        } catch (e: Throwable) {
            _phase.value = WipeRestartPhase.Failed("PRF assertion failed: ${e.message}")
            return
        }

        // 3 — Wrap NEW UMK under PRF secret.
        _phase.value = WipeRestartPhase.WrappingNewUmk
        val sealed = Recovery.wrap(newUmk, prfSecret)
        // The Worker hashes the decoded ciphertext bytes (nonce||ct).
        val nonceBytes = java.util.Base64.getDecoder().decode(sealed.nonceBase64)
        val ctBytes = java.util.Base64.getDecoder().decode(sealed.ciphertextBase64)
        val combined = nonceBytes + ctBytes
        val newWrappedUmkB64 = java.util.Base64.getEncoder().encodeToString(combined)
        val wrappedHashHex = sha256Hex(combined)

        // 4 — Derive NEW IRK pub from NEW UMK at v1.
        _phase.value = WipeRestartPhase.Signing
        val newIrkSeed = Hkdf.computeHkdf(
            /* macAlgorithm = */ "HMACSHA256",
            /* ikm = */ newUmk,
            /* salt = */ "flagship/irk/v1".toByteArray(),
            /* info = */ "ed25519-seed".toByteArray(),
            /* size = */ 32,
        )
        val newPair = Ed25519Sign.KeyPair.newKeyPairFromSeed(newIrkSeed)
        val newIrkPubHex = HexUtil.encode(newPair.publicKey)
        val issuedAt = System.currentTimeMillis()
        val canonical = WipeRestartClaim.canonicalBytes(
            username = user,
            oldIrkPubHex = oldIrkPubHex,
            newIrkPubHex = newIrkPubHex,
            newCredentialIdHex = credentialId,
            newWrappedUmkHashHex = wrappedHashHex,
            issuedAt = issuedAt,
        )
        val signature = oldIrkSign.sign(canonical)

        // 5 — POST.
        _phase.value = WipeRestartPhase.Posting
        val idempotencyKey = randomIdempotencyKey()
        try {
            server.wipeRestart(
                username = user,
                body = WipeRestartRequest(
                    request = WipeRestartRequest.Inner(
                        username = user,
                        oldIrkPub = oldIrkPubHex,
                        newIrkPub = newIrkPubHex,
                        newCredentialId = credentialId,
                        newWrappedUmk = newWrappedUmkB64,
                        issuedAt = issuedAt,
                    ),
                    signature = HexUtil.encode(signature),
                    idempotencyKey = idempotencyKey,
                ),
                ifMatch = currentEtag,
            )
        } catch (e: Throwable) {
            val msg = e.message.orEmpty()
            val friendly = when {
                msg.contains("412") -> "Your device list changed in the background. Refresh and try again."
                msg.contains("429") -> "Wipe rate-limited (1 per hour). Try again later."
                msg.contains("409") -> "Another rotation completed first. Your account is fine — refresh and check Activity for the audit trail."
                else -> "Couldn't reach the server: $msg"
            }
            _phase.value = WipeRestartPhase.Failed(friendly)
            return
        }

        // 6 — Atomic local install.
        _phase.value = WipeRestartPhase.InstallingLocally
        try {
            Keystore.installUmk(newUmk)
            // Drop our push token — it's bound to an identity that
            // no longer exists on .com.
            Keystore.setPushTokenId(null)
        } catch (e: Throwable) {
            _phase.value = WipeRestartPhase.Failed(
                "Server committed but local install failed: ${e.message}. Open the app fresh to recover.",
            )
            return
        }
        _phase.value = WipeRestartPhase.Completed
    }

    private fun sha256Hex(data: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        val h = md.digest(data)
        return h.joinToString("") { "%02x".format(it) }
    }

    private fun randomIdempotencyKey(): String {
        val bytes = ByteArray(16).also(rng::nextBytes)
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/WipeRestartViewModel.swift
import Foundation
import Observation
import CryptoKit
import Flagship
import FlagshipAPI
import FlagshipCore

/// E2 — Wipe & restart orchestrator. Rotates everything the user can
/// rotate, in one atomic ceremony:
///
///   1. Generate a fresh UMK seed locally.
///   2. Run WebAuthn passkey REGISTER → new credentialID.
///   3. PRF-assert against the new credential → 32-byte secret.
///   4. AES-GCM wrap the NEW UMK under the new PRF secret →
///      new wrappedUmk + nonce.
///   5. Compute SHA-256 of the new wrappedUmk so the canonical bytes
///      stay small (the protocol signs the HASH of the ciphertext,
///      not the ciphertext itself).
///   6. Derive OLD IRK (from current UMK) — this device's existing
///      authority. The OLD IRK signs the WipeRestart envelope, which
///      is the trust basis that lets .com swap the registered IRK.
///   7. Derive NEW IRK from the NEW UMK at v1 — what the device will
///      use going forward.
///   8. Sign canonical flagship/wipe-restart/v1 bytes with the OLD IRK.
///   9. POST /api/users/:u/wipe-restart with body + 16-byte
///      idempotencyKey. Server CAS-swaps IRK + envelope + appends
///      audit row atomically.
///  10. On 200: install the NEW UMK into Keystore (which resets the
///      IRK version slot). Local state now matches what .com just
///      committed.
///
/// Failure handling is conservative: until step 10 succeeds, Keystore
/// retains the OLD UMK — so a network blip on the POST doesn't strand
/// the user with mismatched local + server keys.
@Observable
@MainActor
public final class WipeRestartViewModel {

    public enum Phase: Equatable, Sendable {
        case idle
        case preparingKeys
        case registeringPasskey
        case wrappingNewUmk
        case signing
        case posting
        case installingLocally
        case completed
        case failed(String)
    }

    public private(set) var phase: Phase = .idle

    private let server: FlagshipServerClient
    private let webAuthn: WebAuthnProvider
    private let username: () -> String?

    public init(
        server: FlagshipServerClient,
        webAuthn: WebAuthnProvider = MockWebAuthnProvider(),
        username: @escaping () -> String?
    ) {
        self.server = server
        self.webAuthn = webAuthn
        self.username = username
    }

    /// Run the ceremony. `currentEtag` is the value the caller
    /// captured from its most recent `listDevices` call — passed as
    /// `If-Match` to fence the device-list-shifted race. Pass nil
    /// to skip the fence (the rate limit + signature gate still apply).
    public func run(currentEtag: String?) async {
        guard let user = username(), !user.isEmpty else {
            phase = .failed("No active account on this device.")
            return
        }

        // 1 — Materialize OLD + NEW UMKs.
        phase = .preparingKeys
        let oldUmk: SymmetricKey
        do {
            oldUmk = try await Keystore.currentUMK(reason: "Wipe & restart")
        } catch {
            phase = .failed("Couldn't access your current account key: \(error.localizedDescription)")
            return
        }
        let newUmk = SymmetricKey(size: .bits256)

        // 2 — Register a fresh passkey.
        phase = .registeringPasskey
        let registration: WebAuthnRegistration
        do {
            registration = try await webAuthn.register()
        } catch {
            phase = .failed("Passkey registration failed: \(error.localizedDescription)")
            return
        }
        // 3 — PRF assert against the new credential.
        let prfSecret: Data
        do {
            prfSecret = try await webAuthn.prfAssert(credentialId: registration.credentialId)
        } catch {
            phase = .failed("PRF assertion failed: \(error.localizedDescription)")
            return
        }

        // 4 — Wrap the NEW UMK under the new PRF secret.
        phase = .wrappingNewUmk
        let wrapped: (ciphertextBase64: String, nonceBase64: String)
        do {
            wrapped = try Recovery.wrap(umkSeed: newUmk, prfSecret: prfSecret)
        } catch {
            phase = .failed("Couldn't wrap new key: \(error.localizedDescription)")
            return
        }
        // The Worker stores wrappedUmk as a single base64 string that
        // it decodes + hashes. The Worker's canonical bytes also
        // SHA-256 the decoded ciphertext bytes (NOT base64 chars), so
        // we feed it the raw nonce||ct bytes the SealedBox produces.
        guard let nonceBytes = Data(base64Encoded: wrapped.nonceBase64),
              let ctBytes = Data(base64Encoded: wrapped.ciphertextBase64)
        else {
            phase = .failed("Local base64 round-trip failed")
            return
        }
        let combined = nonceBytes + ctBytes
        let newWrappedUmkB64 = combined.base64EncodedString()
        let wrappedHashHex = sha256Hex(combined)

        // 5 — Derive OLD + NEW IRK pub keys.
        phase = .signing
        let oldIrk: Curve25519.Signing.PrivateKey
        let newIrkPubHex: String
        do {
            oldIrk = try await Keystore.deriveIRK(
                reason: "Authorize wipe",
                version: Keystore.currentIrkVersion()
            )
            // The NEW IRK derives from the NEW UMK at v1. We don't
            // persist it until step 7 (post-success install), but we
            // need its pub to put in the WipeRestart envelope.
            let newIrkSeed = HKDF<SHA256>.deriveKey(
                inputKeyMaterial: newUmk,
                info: Data("flagship/irk/v1".utf8),
                outputByteCount: 32,
            )
            let newIrkKey = try Curve25519.Signing.PrivateKey(
                rawRepresentation: newIrkSeed.withUnsafeBytes { Data($0) }
            )
            newIrkPubHex = HexUtil.encode(newIrkKey.publicKey.rawRepresentation)
        } catch {
            phase = .failed("Couldn't derive identity keys: \(error.localizedDescription)")
            return
        }
        let oldIrkPubHex = HexUtil.encode(oldIrk.publicKey.rawRepresentation)
        let issuedAt = Int64(Date().timeIntervalSince1970 * 1000)
        let canonical = WipeRestartClaim.canonicalBytes(
            username: user,
            oldIrkPubHex: oldIrkPubHex,
            newIrkPubHex: newIrkPubHex,
            newCredentialIdHex: registration.credentialId,
            newWrappedUmkHashHex: wrappedHashHex,
            issuedAt: issuedAt,
        )
        let signature: Data
        do {
            signature = try oldIrk.signature(for: canonical)
        } catch {
            phase = .failed("Couldn't sign: \(error.localizedDescription)")
            return
        }

        // 6 — POST.
        phase = .posting
        let idempotencyKey = randomIdempotencyKey()
        do {
            _ = try await server.wipeRestart(
                username: user,
                body: WipeRestartRequest(
                    request: .init(
                        username: user,
                        oldIrkPub: oldIrkPubHex,
                        newIrkPub: newIrkPubHex,
                        newCredentialId: registration.credentialId,
                        newWrappedUmk: newWrappedUmkB64,
                        issuedAt: issuedAt,
                    ),
                    signature: HexUtil.encode(signature),
                    idempotencyKey: idempotencyKey,
                ),
                ifMatch: currentEtag,
            )
        } catch ScreensClientError.http(let status, _) where status == 412 {
            phase = .failed("Your device list changed in the background. Refresh and try again.")
            return
        } catch ScreensClientError.http(let status, _) where status == 429 {
            phase = .failed("Wipe rate-limited (1 per hour). Try again later.")
            return
        } catch ScreensClientError.http(let status, _) where status == 409 {
            phase = .failed("Another rotation completed first. Your account is fine — refresh and check Activity for the audit trail.")
            return
        } catch {
            phase = .failed("Couldn't reach the server: \(error.localizedDescription)")
            return
        }

        // 7 — Atomic local install. From this point the device's
        // Keystore matches what .com just committed.
        phase = .installingLocally
        do {
            try await Keystore.installUMK(newUmk, reason: "Activate new account key")
        } catch {
            // Worst case here: server committed but local install
            // failed. The peer-detection mechanism (E7) will catch
            // it on next foreground via /api/users/:u/devices, and
            // the user will be prompted to sign in again — which
            // is the right recovery path.
            phase = .failed("Server committed but local install failed: \(error.localizedDescription). Open the app fresh to recover.")
            return
        }
        // Bonus on success: forget OUR push token so we don't try
        // to push-relay to a now-invalid identity. Re-registration
        // happens on next launch via PushRegistrar.
        try? Keystore.setPushTokenId(nil)
        phase = .completed
    }

    // MARK: - helpers

    private func sha256Hex(_ data: Data) -> String {
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private func randomIdempotencyKey() -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}
